src/main/java/play/dev/filewatch/SourceModificationWatch.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (C) from 2022 The Play Framework Contributors <https://github.com/playframework>, 2011-2021 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package play.dev.filewatch;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.Comparator;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+public final class SourceModificationWatch {
+
+  private SourceModificationWatch() {}
+
+  private static Set<File> listFiles(Supplier<Iterable<File>> sourcesFinder) {
+    return StreamSupport.stream(sourcesFinder.get().spliterator(), false)
+        .collect(Collectors.toSet());
```

Comment:
Actually, IMHO this was the only real "mistake" you made from converting the Scala code to Java code:
The original Scala argument was `() => Iterator[ScalaFile]` which translates to `Supplier<Iterable<File>> sourcesFinder`. So this method `listFiles` is responsible to call `.get()` to calculate the iterable.